feat: Support for Mark to Base Attachment Positioning (incl. GPOS 4, 9)

rafallyczkowskiadylic opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Added support for Mark to Base Attachment Positioning (for the DFLT script). Feature is enabled by default because it is a part of a proper glyphs positioning. Includes capabilities implementation for the corresponding parsers (GPOS lookup 4 & 9)

Followed the current solution's convention to introduce new features within the font & layout implementation.

- added GPOS `LookupType 4` parser for `MarkBasePosFormat1` support
- added GPOS `LookupType 9` parser for Extension Positioning `SubtableFormat1` support
- added GSUB support for `multiSubstitutionFormat1` 
- introduced the union lookup feature (`getFeaturesLookups`) for the layout to meet the OT specs
- fixed GSUB substitutions algorithm 
- tests coverage for kerning and positioning features
- tests coverage for all supported and not supported lookups of GPOS

**TODO**: Add support for non DFLT scripts for `MARK` (`src/font.js:344`)
**TODO**: Add a support device offsets for anchor table format 3 (`src/parse.js:515`)
**TODO**: Add MarkToMark Attachment positioning (lookup type 6) functionality.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Positioning changes

Fonts like Mandarin or Thai to successfully render text with all glyphs (sometimes 3+) representing characters like `ปั` or `ริ่`
requiring the `LookupType 4` for mark to base attachment positioning to do it properly. 

[Extension subtables support](https://learn.microsoft.com/en-us/typography/opentype/spec/gpos#lookuptype-9-extension-positioning) is **required** to support the subtables that exceeds the 16-bit limits of the various other offsets in the GPOS table - that genuinely are being used eg. for the [Apples SF Pro fonts](https://developer.apple.com/fonts/) and many others.

### Substitution changes

There was an issue with the current algorithm where it was collecting all substitutions from all lookup tables and after that each substitution was conducted. This is invalid. Because it should apply substitution for each lookup table before moving to the next lookup table. In other words: substitution should be applied before moving to the next lookup table 

Specification:
> During text processing, a client applies a lookup to each glyph in the string before moving to the next lookup. A lookup is finished for a glyph after the client locates the target glyph or glyph context and performs a substitution, if specified. To move to the “next” glyph, the client will typically skip all the glyphs that participated in the lookup operation: glyphs that were substituted as well as any other glyphs that formed a context for the operation.

Ref: https://learn.microsoft.com/en-us/typography/opentype/spec/gsub#table-organization

### The union lookup feature

The union lookup feature (`getFeaturesLookups`) was required actually to meet the [OpenType specification](https://learn.microsoft.com/en-us/typography/opentype/otspec191alpha/chapter2#lookup-list-table) which guarantees that all enabled features will be processed in a valid order both for positioning and substitutions.
> the list of lookups is the union of lookups referenced by all of those features, and these are all processed in their lookup list order.


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Added tests coverage to the `test/parser.js` and `test/tables/gpos.js` including new binary data and testing:
- the [Lookup Type 4: Mark-to-Base Attachment Positioning Subtable specification](https://learn.microsoft.com/en-us/typography/opentype/spec/gpos#lookup-type-4-mark-to-base-attachment-positioning-subtable)
- the [LookupType 9: Extension Positioning](https://learn.microsoft.com/en-us/typography/opentype/spec/gpos#lookuptype-9-extension-positioning)

These lookups uses existing structures like coverage tables, and were tested agains different formats.  Extension positioning lookup was tested agains all currently supported and not supported subtable parsers. 

Added tests coverage to the `test/layout.js` to test the method for union lookups (`getFeaturesLookups`).
Added tests coverage for glyph positioning tests in the `test/font.js` including:
- Kerning feature support with kerning table and lookup table support
- `options.kerning` and  `options.features.kern` testing 
- Mark To Base Attachment feature support to work with kern features
- `options.features.mark` testing
Added tests coverage to the `test/bidi.js` to test `multiSubstitutionFormat1` glyphs decomposition 

Moreover there were many test conducted for the rendered content using `latn` and `arab` fonts.

## Screenshots (if appropriate):

### Positioning issues

BEFORE (INVALID glyphs pos):
![image](https://user-images.githubusercontent.com/102729629/216976249-278734ce-ee78-491c-a19c-8208e730d516.png)

AFTER (FIXED):
![image](https://user-images.githubusercontent.com/102729629/216973120-aad3661c-f59f-4139-a3f3-82bfb9df960d.png)


### Substitution issues

BEFORE (INVALID) - very end glyph is a single glyph that should be decomposed into two.
![image](https://github.com/opentypejs/opentype.js/assets/102729629/a76570ec-5646-408c-acc1-9edd301bc626)

AFTER (FIXED) - decomposed into two glyph: 
![image](https://github.com/opentypejs/opentype.js/assets/102729629/71ec4ab4-4c5c-4f28-bfde-ac4aeaac7d00)

<img width="849" alt="image" src="https://github.com/opentypejs/opentype.js/assets/102729629/17d6b385-8003-4ff7-b4b1-8878300ea2b1">



## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I did `npm run test` and all tests passed green (including code styling checks).
- [x] I have added tests to cover my changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the **README** accordingly.
- [x] I have read the **CONTRIBUTING** document.

rafallyczkowskiadylic committed:
feat: additional GPOS lookups parsers

- added GPOS LookupType 4 parser for MarkBasePosFormat1 support
- added GPOS LookupType 9 parser for Extension Positioning SubtableFormat1 support
- tests coverage for all supported and not supported lookups of GPOS

TODO: Add a support device offsets for anchor (Parser.anchor) table format 3.

rafallyczkowskiadylic committed:
feat: introduced glyphs positioning features (kern, mark)

- introduced an union, ordered features lookup list
- refactored the Font options features update method
- supported kerning table
- supported kern lookup table
- supported mark to base attachments

rafallyczkowskiadylic committed:
fix: code style fixes.

rafallyczkowskiadylic committed:
fix: kerning pos fix for pair

rafallyczkowskiadylic committed:
fix: kern processing value to support entire sequence

- fix for kern sequence
- tests coverage

Connum commented:
Does this make #491 redundant, or do we have to consolidate the two? Could you please have a look at each other's code @rafallyczkowskiadylic & @AlexandreRivet?

ILOVEPIE commented:
@rafallyczkowskiadylic I noticed that rlig is missing an implementation for latin scripts and liga is missing an implementation for arabic scripts, is that also fixed in this PR?

rafallyczkowskiadylic commented:
@Connum https://github.com/opentypejs/opentype.js/pull/491 is related, but not redundant. These MRs have different features covered. But esentially both have a required `rlig` missed :-) 

My proposition for modus operandi:
Let's focus on https://github.com/opentypejs/opentype.js/pull/491 in first place. I will commit and cover the missing `rlig` over there. By the time I will be working on that please review this MR and drop questions or any concerns so we can discuss. 

@ILOVEPIE These two MRs are considered as separate chunks of features. I will fix whatever will be needed starting with the smaller one https://github.com/opentypejs/opentype.js/pull/491

ILOVEPIE reviewed (changes requested):
I could only find one issue with the code besides some naming choices for constants. I'm going to get @Connum to review this as well as it's a larger PR and I want to make sure that it's ok.

ILOVEPIE commented:
Also, @rafallyczkowskiadylic does this PR fix #501, because if you're redoing this code anyways, it might be worth fixing it.

Connum commented on src/substitution.js:
```diff
@@ -12,7 +12,10 @@ import Layout from './layout';
-    Layout.call(this, font, 'gsub');
+    Layout.call(this, font, 'gsub', [
+        { featureName: 'rlig',  supportedLookups: [4] }
+        // TODO: Define all supported features to use layout.getFeaturesLookups for a sequence ordered feature lookups
```
Is this a TODO for some separate future PR, or did you forget to implement this? How important is this, shouldn't it be in this PR?

rafallyczkowskiadylic replied:
@Connum this configuration param is used by the `Layout.getFeaturesLookups` that implements the ordered features lookup sequence. The thing is I have improved only Positioning features lookup (`Position.prototype.getPositionFeatures`) leaving substitutions untouched. So for the substitutions still uses the  "old features lookups parsing flow" which is as it was before. Adapting this ordered lookup sequence to the substitution will require a decent refactoring (and then this line will be "in use"). 

Hope this make sense?

rafallyczkowskiadylic replied:
PS. and just to be on the same page, this array defines which lookup tables are supported for which features. This way we can have less redundancy and a single algorithm that parses the lookup tables. That why `Layout.getFeaturesLookups` got implemented.

We might get refactor **substitutions** with better tests coverage. Right now I have not taken this call because I had some concerns about the tests coverage not mature enough for such change to the substitutions (bidi)

Connum commented on src/layout.js:
```diff
@@ -194,6 +195,61 @@ Layout.prototype = {
+            return [];
+        }
+
+        // Fitler out only supported by layout table features
```
just y small typo, Fitler => Filter

rafallyczkowskiadylic replied:
Fixed.

Connum commented on src/parse.js:
```diff
@@ -477,6 +477,74 @@ Parser.prototype.parseCoverage = function() {
+                anchorPoint: this.parseUShort()
+            };
+
+        // TODO: Add a support Device offsets
```
As with the other TODO, has this been accidentally forgotten, or why is it not implemented with this PR? (not supposed to sound like an accusation/demand, just would like to know what's the reason)

rafallyczkowskiadylic replied:
@Connum it was intended, because actually this was not required to be supported within the iteration. I covered the planned scope which was not including this bit. If you are concerned, I think I might revisit this

rafallyczkowskiadylic replied:
Assuming this comment https://github.com/opentypejs/opentype.js/pull/557#issuecomment-1453904466 was referring to this?:) Please resolve thread if we are on the same page. Thanks.

Connum reviewed (changes requested):
I have to admit that all that byte-level-parsing stuff is a bit out of my comfort zone (yes, despite me having implemented a few parsing features before ;)), but other than @ILOVEPIE's comments and those I added, there's nothing that caught my eye.

Connum commented:
@rafallyczkowskiadylic the feature order is quite a critical thing to add before we start working on implementing more features like `calt` etc. It would be great if we could merge this soon!

rafallyczkowskiadylic commented on src/parse.js:
```diff
@@ -477,6 +477,74 @@ Parser.prototype.parseCoverage = function() {
+                xCoordinate: this.parseShort(),
+                yCoordinate: this.parseShort(),
+                xDevice: 0x00,
+                yDevice: 0x00,
```
Good catch! I will fix shortly

rafallyczkowskiadylic replied:
Fixed. Pointer is moving now

rafallyczkowskiadylic commented:
@Connum once we have the other [MR](https://github.com/opentypejs/opentype.js/pull/491/files) merged (that covers `rlig` that is missing here as well) I will resolve and address all concerns around this mr shortly.

Connum commented:
Sure, that's fine! We don't have to support everything at once when implementing a new feature or format. We should just make sure that users are notified about any missing/unsupported functionality.

I just wanted to avoid that anything was misusing that was supposed to be in this PR already.

rafallyczkowskiadylic commented:
@Connum Perfect! Great you are on top. So, let's address what is pending now. We stay in touch

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mtbase-merge-testing

# Conflicts:
#	src/font.js
#	test/layout.js

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mark-to-base-attachment-positioning

rafallyczkowskiadylic commented on src/font.js:
```diff
@@ -279,7 +282,8 @@ Font.prototype.defaultRenderOptions = {
         { script: 'arab', tags: ['init', 'medi', 'fina', 'rlig'] },
-        { script: 'latn', tags: ['liga', 'rlig'] }
+        { script: 'latn', tags: ['liga', 'rlig'] },
+        { script: 'DFLT', tags: ['mark'] },
```
@ILOVEPIE I could not find this. I believe you might have referred to the langSysTable (tag) and this is a script tag? Moreover:
- when looking into specs,[ there are refs to the uppercased values only](https://learn.microsoft.com/en-us/typography/opentype/otspec170/scripttags) or [here](https://learn.microsoft.com/en-us/typography/opentype/spec/chapter2#script-list-table-and-script-record) (screen) 
- when verified with the font source definitions, found also uppercased (screen) 

![image](https://user-images.githubusercontent.com/102729629/223085009-d78036d2-49bc-4c5d-b054-c48633e1b06c.png)

Connum replied:
https://learn.microsoft.com/en-us/typography/opentype/spec/scripttags it's lowercase for all other tags, but DFLT is uppercase in the docs

rafallyczkowskiadylic committed:
fix: pointer offset fix, typos

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mark-to-base-attachment-positioning

# Conflicts:
#	src/font.js

rafallyczkowskiadylic committed:
fix: tests after merge

rafallyczkowskiadylic commented:
I will check https://github.com/opentypejs/opentype.js/issues/501

rafallyczkowskiadylic committed:
Merge branch 'master' into feature/mark-to-base-attachment-positioning

rafallyczkowskiadylic committed:
fix: fixed lint code after merge

rafallyczkowskiadylic commented:
#501 is fixed within this MR

rafallyczkowskiadylic committed:
fix: Fixes the substitution algorithm and support for multiSubstitutionFormat1

- substitution algorithm to match specficiation (lookup - apply - lookup - apply)
- added support for multiSubstitutionFormat1 substitutions
- updated applySubstitution method for more flexibility
- tests coverage

rafallyczkowskiadylic committed:
fix: context condition check

rafallyczkowskiadylic committed:
fix: code linting and comments

rafallyczkowskiadylic committed:
Merge pull request #5 from Adylic/bugfix/substitution-algorithm-fix

Bugfix/substitution algorithm fix

rafallyczkowskiadylic commented:
@Connum @ILOVEPIE added few more commits with the substitution algorithm fixes and a new substitution format support. I haven't refactored the other working code (just the required bits). We can find few spots where we see "similar" functionalities. For example:

1. `src/features/featureQuery.js`  `FeatureQuery.prototype.lookupFeature` collects substitutions - too much responsibility here. I marked this as deprecated and pointed the fixed algorithm
2. `src/substitution.js` `Substitution.prototype.getFeature` - substitution processing
3. `src/bidi.js` - substitution applications 

I believe we could move towards:
- `Layout` - to keep what it has, consume GPOS/GSUB definitions, lookups. Keep important `getFeaturesLookups` method that build a union lookups list for ordered processing.
- `Position` and `Substitution` - shares base functionality to provide features lookups through the `Layout`
- `FeatureQuery` - should be promoted to use this to find and process features (not only in the bidi)
- substitution features per script i.e `src/features/latn/latinLigatures.js` `src/features/arab/arabicRequiredLigatures.js` should not be required since mostly they are simply the same. We basically need word ranges and context, but these are the same methods. These can be removed in favor of  `applySubstitutions.call(this, 'thai', ['liga', 'rlig', 'ccmp']);`

Not sure if `bidi` is really required, maybe for the directional substitutions but I think bidi could be deprecated and it's responsibility could be dropped into the `Font` where bidi is being called and where positioning logic is happening anyways. But basically I think we could use a  single processor  both for substitutions and positioning. 

Well this can open a discussion, but wanted to share some thoughts.

Connum commented:
I'm totally for unifying duplicate functionality. The way that feature processing is currently structured is something that has deterred me from working on new font features because it frankly seems a bit messy and not very intuitive to me. However, I would see that as a separate step from this PR. Let's get this support in first and consolidate feature processing in the future. Otherwise there might be too much code to review at once, and we have enough open and partly almost-finished PRs lying around as it is.

That's my opinion and not the final say, of course.

ILOVEPIE commented:
This is blocking quite a few PRs, we should probably get this merged sometime soon.

rafallyczkowskiadylic commented:
Thank you guys for the comments. Yes, most of suggestions/work has been done.  I have consolidated functionality as described introducing more generic functions to handle/fix specifications. I will resolve conflicts  soon.

Connum commented:
Fixes #501 according to https://github.com/opentypejs/opentype.js/issues/501#issuecomment-1543727128

@rafallyczkowskiadylic it's been a couple of weeks - just a friendly nudge. :)

Connum commented on src/layout.js:
```diff
@@ -194,6 +195,61 @@ Layout.prototype = {
+     * https://learn.microsoft.com/en-us/typography/opentype/otspec191alpha/chapter2#lookup-list-table
+     *
+     * @param {string[]} requestedFeatures
+     * @param {string} [script='DFLT']
```
according to the spec it's uppercase: https://learn.microsoft.com/en-us/typography/opentype/spec/scripttags

Connum commented:
I was finally able to test this more thoroughly... I merged it into my experimental fork after resolving some conflicts.

The chars in the GPOS–3 test of the unicode test suite still rendered wrong, and they also didn't work in our demo page. 
![image](https://github.com/opentypejs/opentype.js/assets/13076806/56326829-c9f1-4ef1-8567-308254c4cc80) 
![image](https://github.com/opentypejs/opentype.js/assets/13076806/a7167bd3-6fde-40d9-9168-09b71b918846)


So I had to make a few changes to make it work, see
https://github.com/Connum/opentype.js-next/commit/95688c6f54ec55274cfe2eee86eb8493788ca632

![image](https://github.com/opentypejs/opentype.js/assets/13076806/ef0b92c7-4a12-4406-90ae-71251e5b07e5)
![image](https://github.com/opentypejs/opentype.js/assets/13076806/731792bf-a8e6-496a-93dd-6184d9d1e298)
Hey presto!

However, these calculations are quite expensive, even before my additions, because they are done for each single glyph. I don't think that's necessary and we should implement ways to at least cache this, or best only call the feature update when something has changed.

Connum commented:
Ok, now the tests for disabling the mark feature are failing of course, because I implemented it so that default features cannot be disabled. We'll have to talk about how we want to handle this.

Connum commented:
The more I look at this, the more confused I get... 😅

I think we still need some cleaning up before we can finally merge this, but I'm lost here without you @rafallyczkowskiadylic. I tried to fix https://rawgit.com/unicode-org/text-rendering-tests/master/reports/OpenType.js.html#GSUB-1, but I got caught up in spaghetti code for several hours trying to follow the path of the substitution mechanism and got confused between old and new functionality. e.g. we should really get rid of this:
> - substitution features per script i.e `src/features/latn/latinLigatures.js` `src/features/arab/arabicRequiredLigatures.js` should not be required since mostly they are simply the same. We basically need word ranges and context, but these are the same methods. These can be removed in favor of  `applySubstitutions.call(this, 'thai', ['liga', 'rlig', 'ccmp']);`

Regarding this substitution case, I started wondering why we split into words in the first place? The space glyph being part of a substitution is perfectly fine, there's no rule against it to my knowledge.

I know that's the crux of open source, but I wouldn't be such a pest if it wasn't blocking the progress so much. A clear structure for substitution and positioning as well as supporting all the different formats is really crucial for implementing further features.

Connum commented:
By the way, trying to get rid of the latn-specific `applyLatinLigatures`, I replaced it with
```
function applyLatinFeatures() {
    checkGlyphIndexStatus.call(this);
    applySubstitutions.call(this, 'latn', ['liga', 'rlig', 'ccmp', 'calt']);
}
```
in analogy to `applyThaiFeatures()`, and called it in `Bidi.prototype.applyFeaturesToContexts`:
```
    if (this.checkContextReady('latinWord')) {
        applyLatinFeatures.call(this);
    }
```

But then all ligatures stopped working...

rafallyczkowskiadylic commented:
Hey @Connum  I am on it. Let me resolve conflicts and see what is on the code. I will revert back by the end of the weekend

rafallyczkowskiadylic committed:
fix: Resolved conflicts

rafallyczkowskiadylic committed:
Merge pull request #6 from Adylic/resolve-conflicts

Resolve conflicts

rafallyczkowskiadylic commented:
@Connum Conflicts resolved. 

Pending items:

1. Could you share this `TestShapeEthi.ttf` file with me so I can verify the outcomes? 
2. I understand. I had similar issue, with some redundancy to get into the logic, that is why I didn't wanted to deprecate old code but rather leverage a new core lookup features. The most important functions is: Layout.getFeaturesLookups that implements common logic for fetching features, for substitutions and for positioning. 

> I started wondering why we split into words in the first place?

It was implemented like that, and the reason for this is that within single sentence there may be characters from a different language context. Example: `我的iPhone是最好的` where "iPhone" is for the latin context word. 

4. For the `applyLatinLigatures` issue, could you send me specific case you are testing? Including exact string, expected result, and source font file. I will check what may be happening because I see no issues with the latin fonts I am testing currently. 

Moreover, please keep in mind supported substitution formats.

```
/**
 * Supported substitutions
 */
const SUBSTITUTIONS = {
    11: singleSubstitutionFormat1,
    12: singleSubstitutionFormat2,
    63: chainingSubstitutionFormat3,
    41: ligatureSubstitutionFormat1, 
    21: multiSubstitutionFormat1
};
```

Different fonts' source files may have these definitions in a different formats that eventually may not be supported yet.

-----

@Connum if you could share all your use cases, with results you expect to see and source font files - I can review all of them and eventually address. 

-----

For the clean up I think we could just mark functions as deprecated for now. I'd refactor only bits covered with reliable tests so we can actually rely on outcomes before/after we change the code (this was also the reason I'd rather applied the code and "promoted" than removing or refactoring old functioning pieces).

rafallyczkowskiadylic commented:
Is there easy way to tests svg using demo page? I wrapped up simple compiler so I am using this a

`npm run font:preview -- --font="Roboto-Black.ttf" --content='Lorem ipsum dolor sit amet. ąłćżółźżćśćłóęą '`

that generates `preview.html` that compares HTML (browser) rendered content and SVG path. 

![image](https://github.com/opentypejs/opentype.js/assets/102729629/6cd586de-ece1-47d9-b545-925e99bb8d03)

rafallyczkowskiadylic commented:
Ok. @Connum currently I do not have pending items, unless you share your failing test cases (font names, conetnt tested, etc) with me according [to this](https://github.com/opentypejs/opentype.js/pull/557#issuecomment-1837164457)

Connum commented:
I don't really like the idea of having both approaches in the code base in parallel. The arab and latn features should work with the new approach as well, so we should move everything to work with it. I understand if this is more than you're willing to do. Could you mark the PR as allowing maintainers to make changes? Then me or someone else could look into fixing this for all the scripts.

Connum commented:
Shouldn't I be able to change `applyLatinLigatures()` to
```
function applyLatinLigatures() {
    applySubstitutions.call(this, 'latn', ['liga']);
}
```
and it should just work? But the ligature tests then fail.

rafallyczkowskiadylic committed:
fix: explicit tokenizer context name

- script: latn, was not matching latinWord

rafallyczkowskiadylic commented:
> Shouldn't I be able to change `applyLatinLigatures()` to
> 
> ```
> function applyLatinLigatures() {
>     applySubstitutions.call(this, 'latn', ['liga']);
> }
> ```
> 
> and it should just work? But the ligature tests then fail.

Actually I checked and found few tests failing for your cases, for `latn` they are related to manipulating font's definitions (`Substitution.prototype.addLigature`). These implementations are still using old lookupTables search, and manipulating local references (

In other word, addLigature is using `Layout.getLookupTables`, manipulates this method's output references (expecting the test will also use `.getLookupTables` method). Thing is my solution does not use `.getLookupTables` at all because of `getFeaturesLookups` 

```
Substitution.prototype.addLigature = function(feature, ligature, script, language) {
    const lookupTable = this.getLookupTables(script, language, feature, 4, true)[0];
    let subtable = lookupTable.subtables[0];
    ...
    subtable.ligatureSets = .... // <- manipulation
});
```

yet `Layout.getLookupTables` creates own copy of subtables refs (that is why we can't see them within new feature) 

![image](https://github.com/opentypejs/opentype.js/assets/102729629/40902d14-c5c6-456d-a8ef-23e1bd2b446e)

If `.addLigrature` would instead manipulate definition on the `this.font.tables[this.tableName].lookups` level (not local references) everything will work as expected. This is how I would approach this, to deprecate any use of `getLookupTables` and if there are definitions updated I would try to make them as "high" as possible (ideally around specific table def source, `this.font.tables[this.tableName]....`)
